src/icrc2-minter/src/ops.rs

```diff
@@ -0,0 +1,421 @@
+use bridge_canister::bridge::{Operation, OperationAction, OperationContext};
+use bridge_did::error::{BftResult, Error};
+use bridge_did::id256::Id256;
+use bridge_did::op_id::OperationId;
+use bridge_did::order::{self, MintOrder, SignedMintOrder};
+use bridge_did::reason::Icrc2Burn;
+use bridge_utils::bft_events::{BurntEventData, MintedEventData, NotifyMinterEventData};
+use bridge_utils::evm_link::address_to_icrc_subaccount;
+use candid::{CandidType, Decode, Nat};
+use did::{H160, H256, U256};
+use ic_exports::ic_kit::RejectionCode;
+use ic_task_scheduler::retry::BackoffPolicy;
+use ic_task_scheduler::task::TaskOptions;
+use icrc_client::account::Account;
+use icrc_client::transfer::TransferError;
+use serde::{Deserialize, Serialize};
+
+use crate::constant::IC_CHAIN_ID;
+use crate::tokens::icrc1::{self, IcrcCanisterError};
+use crate::tokens::icrc2::{self, Success};
+
+#[derive(Debug, Serialize, Deserialize, CandidType, Clone)]
+pub enum IcrcBridgeOp {
+    // Deposit operations:
+    BurnIcrc2Tokens(Icrc2Burn),
+    SignMintOrder {
+        order: MintOrder,
+        is_refund: bool,
+    },
+    SendMintTransaction {
+        src_token: Id256,
+        dst_address: H160,
+        order: SignedMintOrder,
+        is_refund: bool,
+    },
+    ConfirmMint {
+        src_token: Id256,
+        dst_address: H160,
+        order: SignedMintOrder,
+        tx_hash: Option<H256>,
+        is_refund: bool,
+    },
+    WrappedTokenMintConfirmed(MintedEventData),
+
+    // Withdraw operations:
+    MintIcrcTokens(BurntEventData),
+    IcrcMintConfirmed {
+        src_address: H160,
+        icrc_tx_id: Nat,
+    },
+}
+
+impl IcrcBridgeOp {
+    pub fn get_signed_mint_order(&self, token: &Id256) -> Option<SignedMintOrder> {
+        match self {
+            Self::SendMintTransaction {
+                order, src_token, ..
+            } if src_token == token => Some(*order),
+            Self::ConfirmMint {
+                order, src_token, ..
+            } if src_token == token => Some(*order),
+            _ => None,
+        }
+    }
+}
+
+impl Operation for IcrcBridgeOp {
+    async fn progress(self, id: OperationId, ctx: impl OperationContext) -> BftResult<Self> {
+        match self {
+            IcrcBridgeOp::BurnIcrc2Tokens(burn_info) => {
+                Self::burn_icrc_tokens(ctx, burn_info, id.nonce()).await
+            }
+            IcrcBridgeOp::SignMintOrder { order, is_refund } => {
+                Self::sign_mint_order(ctx, order, is_refund).await
+            }
+            IcrcBridgeOp::SendMintTransaction {
+                order,
+                src_token,
+                dst_address,
+                is_refund,
+            } => Self::send_mint_tx(ctx, order, src_token, dst_address, is_refund).await,
+            IcrcBridgeOp::ConfirmMint { .. } => {
+                log::warn!("ConfirmMint task should progress only on the Minted EVM event");
```

Comment:
Let's just log all errors in the generic bridge, and not worry about logging in the concrete implementations.

Reply:
Agree